parameter_expansion/pe.py

```diff
@@ -49,22 +105,55 @@ class ParameterExpansionParseError(Exception):
     pass
 
 
-def expand_tokens(s, env):
-    shl = shlex(s, posix=True)
-    shl.commenters = ""
+def expand_tokens(s, env, strict=False):
+    tokens = tokenize(s)
     while True:
         try:
-            yield "".join(takewhile(lambda t: t != "$", shl))
-            yield follow_sigil(shl, env)
+            before_dollar = "".join(takewhile(lambda t: t != "$", tokens))
+            logger_debug("expand_tokens: before_dollar:", repr(before_dollar))
+            yield before_dollar
+            sigil = follow_sigil(tokens, env, strict)
+            logger_debug("expand_tokens: sigil:", repr(sigil))
+            yield sigil
         except StopIteration:
             return
 
 
-def follow_sigil(shl, env):
+def tokenize(s):
+    """Yield token strings lexed from the shell string s."""
+    shl = shlex(s, posix=True)
+    shl.commenters = ""
+    shl.whitespace = ""
+    for grouped, group in groupby(shl, key=is_whitespace):
+        # we group contiguous whitespaces in one string
+        if grouped:
+            yield "".join(group)
+        else:
+            yield from group
+
+
+_get_non_nested_expressions = re.compile(r"(\$\{[^${}]+\})").findall
+
+
+def get_plain_expressions(s):
+    """Return a list of plain, non-nested shell expressions found in the shell
+    string s. These are shell expressions that do not further contain a nested
+    expression and can therefore be resolved indenpendently.
+
+    For example::
+    >>> get_plain_expressions("${_pyname%${_pyname#?}}")
+    ['${_pyname#?}']
+    """
+    return _get_non_nested_expressions(s)
+
+
+def follow_sigil(shl, env, strict=False):
     param = next(shl)
     if param == "{":
-        consume = iter(list(takewhile(lambda t: t != "}", shl)))
-        return follow_brace(consume, env)
+        consume = list(takewhile(lambda t: t != "}", shl))
+        logger_debug("follow_sigil: consume:", consume)
+        consume = iter(consume)
```

Comment:
I guess you convert it to a list to make it possible to log it, but then converting it back to an iterable seems counter-productive. Maybe it'd be better to only convert it to a list if `TRACE` is active.

Reply:
Sure thing. The previous code was already doing iter(list(iter somehow in https://github.com/kojiromike/parameter-expansion/blob/3efeb4e7fbcc10daab6590cc16e06a674474158e/parameter_expansion/pe.py#L66 but that's not a reason to fix this

Reply:
so a better approach may be:
```Python
        consume = takewhile(lambda t: t != "}", shl)
        if TRACE:
            consume = list(consume)
            logger_debug("follow_sigil: consume:", consume)
            consume = iter(consume)
```

... unless there was a reason at first to consume the iterator?

Reply:
@kojiromike actually consuming the iterator is required or else tests fail. Inserting the logging call in between is just a convenience, but consuming (e.g. warping in a list) is not done for the purpose of logging at all.
I added a few comments to clarify this.